### PR TITLE
[FIX][fields] Make [(5,)] with computed domain work.

### PR DESCRIPTION
**Pull request by yajo**
Description of the issue/feature this PR addresses: It makes `[(5,)]` calls in one2many fields work fine when the domain is callable.

Current behavior before PR: `self` in the domain method became the comodel instead of the model when doing action `5`. v9 is affected too, and it gets worse because when doing `record.one2manyfield_ids = False`, v9 converts that to a `5` call (v8 makes it a `6` one, weird... :confused:), which makes computed-domain models crazy.

Desired behavior after PR is merged: When you perform a `5` call inside a one2many, it will effectively evaluate the domain with `self` being the original model, and not the comodel, so you can use things such as other methods in `self` or properties like `self._name`.
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr
@Tecnativa

### CODE REVIEW

File: openerp/addons/test_new_api/tests/test_new_fields.py [hunk 1]
```diff
@@ -468,8 +468,14 @@ def test_60_x2many_domain(self):
         self.assertNotIn(message, discussion.important_messages)
 
         message.important = True
+        discussion.invalidate_cache()
```
Comment (Yajo) [inline openerp/addons/test_new_api/tests/test_new_fields.py]: I had to add this line because somehow it seems like cache for models that have a computed domain do not update cache correctly. If you tell me how to fix that, I can leverage this PR to fix that too, but honestly it's not so much critical as the issue I'm addressing here.

File: openerp/addons/test_new_api/models.py [hunk 2]
```diff
@@ -143,12 +143,16 @@ class Discussion(models.Model):
...
-    important_messages = fields.One2many('test_new_api.message', 'discussion',
-                                         domain=[('important', '=', True)])
+    important_messages = fields.One2many(
+        'test_new_api.message', 'discussion',
+        domain=lambda self: self._domain_important())
     emails = fields.One2many('test_new_api.emailmessage', 'discussion')
     important_emails = fields.One2many('test_new_api.emailmessage', 'discussion',
                                        domain=[('important', '=', True)])
 
+    def _domain_important(self):
```
Comment (pedrobaeza) [inline openerp/addons/test_new_api/models.py]: Maybe you should indicate that the domain is put this way for trying the error or develop another separate test for this.

**Comment by pedrobaeza** [comment]
@rco-odoo @odony Please, this one is important to fix to solve a lot of problems with the use of the new API, and affects all versions.

**Comment by Yajo** [comment]
BTW I split this in several commits so you can simply chekcout b3c5bef and watch the `AttributeError` with your own eyes.

**Comment by rafaelbn** [comment]
:+1:

**Comment by Yajo** [comment]
Do you know another funny thing?

Imagine in this case you have something like `important_messages = test_new_api.messages(2, 3, 4)`. Then you want to remove `2` and add a new one:

```
messages = discussion.important_messages[1:]
messages |= messages.new({...})
discussion.important_messages = messages
```

Odoo v9 will call write with these `vals`:

```
{"important_messages": [
  (5),
  (4, 3),
  (4, 4),
  (0, 0, {...}),
]}
```

So, now that the domain actually works, and the `(5)` part actually removes what it should remove, you get `MissingError` when it reaches `(4, 3)`. *[facepalm]* Preparing test & fix for that too.

Still not tested in v8, but I guess it issues a `(6, 0, [3, 4]), (0, 0, {...})`, which according to docs is equivalent to 5+4+0, so I guess it will fail equally.

I think **this is one huge bug** in the framework, so I'd appreciate some comments.

Not sure, but this could be related to #13082.

File: openerp/addons/test_new_api/models.py (hunk 2 above)
Comment (Yajo) [inline openerp/addons/test_new_api/models.py]: Done.

**Comment by Yajo** [comment]
I reviewed master and it seems the old api removal refactoring fixed it there.

File: openerp/addons/test_new_api/tests/test_new_fields.py (hunk 1 above)
Comment (Yajo) [inline openerp/addons/test_new_api/tests/test_new_fields.py]: I bet I'll be able to remove this after merging #13082

File: openerp/addons/test_new_api/models.py [hunk 3]
```diff
@@ -143,12 +143,17 @@ class Discussion(models.Model):
...
+        """Ensure computed O2M domains work as expected."""
+        return [("important", "=", True)]
+
```
Comment (rco-odoo) [inline openerp/addons/test_new_api/models.py]: Can you please revert the changes on field `important_messages`, and create a new field for this case? Your modification actually discards the original test: invalidating the cache before checking its consistency!

I suggest another, equivalent, field like
```python
# ensure that computed o2m domains work as expected
very_important_messages = fields.One2many('test_new_api.message', 'discussion',
                                          domain=lambda self: self._domain_very_important())

def _domain_very_important(self):
    return [('important', '=', True)]
```

File: openerp/addons/test_new_api/tests/test_new_fields.py [hunk 4]
```diff
@@ -468,8 +468,14 @@ def test_60_x2many_domain(self):
...
         self.assertIn(message, discussion.important_messages)
 
+        discussion.write({
+            "important_messages": [(5,)],
+        })
+        self.assertFalse(discussion.important_messages)
+
```
Comment (rco-odoo) [inline openerp/addons/test_new_api/tests/test_new_fields.py]: Please restore the original test, and replace that code by one on the field `very_important_messages`:
```python
# writing on very_important_messages should call its domain method
self.assertIn(message, discussion.very_important_messages)
discussion.write({'very_important_messages': [(5,)]})
self.assertFalse(discussion.very_important_messages)
self.assertFalse(message.exists())
```

**Comment by Yajo** [comment]
Changes done, thanks, it looks much better now.

Although I'm not sure if tests will pass until https://github.com/odoo/odoo/pull/13082 gets in...
